oauth2-authorization-server/src/main/java/org/springframework/security/oauth2/server/authorization/client/JdbcRegisteredClientRepository.java

```diff
@@ -150,9 +150,15 @@ public void setRegisteredClientParametersMapper(Function<RegisteredClient, List<
 	@Override
 	public void save(RegisteredClient registeredClient) {
 		Assert.notNull(registeredClient, "registeredClient cannot be null");
-		RegisteredClient foundClient = this.findBy("id = ? OR client_id = ? OR client_secret = ?",
-				registeredClient.getId(), registeredClient.getClientId(),
-				registeredClient.getClientSecret().getBytes(StandardCharsets.UTF_8));
+		RegisteredClient foundClient;
+		if (registeredClient.getClientSecret() != null) {
+			foundClient = this.findBy("id = ? OR client_id = ? OR client_secret = ?",
```

Comment:
`RegisteredClient.id` must be unique so the query could simply use `id`.

Also, `RegisteredClient.clientId` must also be unique so you can use either of the two.